### ISSUE TITLE
Improving the transforms

### ISSUE THREAD

**Issue by hubertjb** [braindecode/braindecode#72]
The most recent implementation of transforms asks for an OrderedDict of callables or mne.Raw/Epochs methods that are called successively on the internal Raw/Epochs objects in a ConcatDataset (https://github.com/braindecode/braindecode/blob/master/braindecode/datautil/transforms.py#L18). I wanted to discuss possible improvements to this design:

1. The current design requires the callables or methods to act in-place, i.e., they must directly modify the Raw/Epochs object. However, apart from `.resample` and `.filter,` most mne methods do not seem to act in-place. As for callables, they can be made to modify the objects in-place, but this would mean all the transforms would have to be mne object-aware. I don't have a specific solution in mind, but maybe we could discuss this here.
2. There is currently no way to apply a transform on-the-fly, i.e., every time a window is about to be returned. This is central to preprocessing steps in a lazy loading scenario however. The design we had started to implement during the sprint used Transform objects that were saved in the dataset object, and called inside `__getitem__`.  This is how torch-vision does it for instance.

What are your thoughts @gemeinl @sbbrandt ?

**Comment by gemeinl** [comment]
1. Requiring mne object-awareness seems fine to me. We rely on mne a lot and hence also expect users to know mne and the mne API. We sould also keep <code>apply_function()</code> (https://mne.tools/stable/generated/mne.io.Raw.html#mne.io.Raw.apply_function) in mind for custom transforms. 

2. I think, we first need to know whether on-the-fly transforms are fast enough to be used with lazy loading. Would storing transforms in the datasets then be in addition to the current transforms?

**Comment by hubertjb** [comment]
>     Requiring mne object-awareness seems fine to me. We rely on mne a lot and hence also expect users to know mne and the mne API. We sould also keep `apply_function()` (https://mne.tools/stable/generated/mne.io.Raw.html#mne.io.Raw.apply_function) in mind for custom transforms.

True, `apply_function` would help here, and it conveniently works in-place! We could rewrite `_transform` then:
```python
def _transform(raw_or_epochs, transforms):
    for transform, transform_kwargs in transforms.items():
        if callable(transform):
            raw_or_epochs.apply_function(transform, transform_kwargs)
        else:
            if not hasattr(raw_or_epochs, transform):
                raise AttributeError(
                    f'MNE object does not have {transform} method.')
            getattr(raw_or_epochs.load_data(), transform)(**transform_kwargs)
```

>     I think, we first need to know whether on-the-fly transforms are fast enough to be used with lazy loading.

I think there is no way around this with super large datasets, don't you think? Unless we want to keep preprocessing everything offline, saving the preprocessed files to disk, and then loading them at training time.

Another point here is that if we were to work directly on numpy arrays (instead of on MNE objects) we might get a little performance gain with some transforms.

> Would storing transforms in the datasets then be in addition to the current transforms?

That was part of my question - the way we had implemented it during the sprint was the transforms were stored in the dataset, and either applied in the `__init__` (if loading == eager) or in `__getitem__` (if loading == lazy). The new `transform_concat_ds` function takes care of the first case, but outside of the dataset object. Do we need both approaches? I liked the solution with a list of transforms stored inside the dataset object as it's pretty flexible and also follows torch-vision. The `transform_concat_ds` function is nice too though, but it limits us to the eager loading scenario I think.

**Comment by robintibor** [comment]
I would argue for first having a clean and simple implementation of the eager/already preprocessed lazy loading version. From that we can still extend to on the fly, which we may only need on per-crop basis in the end anyways. Better to have clean fully implemented version that runs on actual examples and going from there.

**Comment by robintibor** [comment]
ok no we have a working version for eager case. I still would like to make some small improvements to the usage, e.g., right now we supply an OrderedDict, while we should supply simply a list. We don't use the Dict functionality as far as I see(?) and we already created a bug/unintended behavior in an example!
https://github.com/braindecode/braindecode/blob/914d1c9961a6747610e206fed84393f118cab706/examples/plot_bcic_iv_2a_moabb.py#L83-L88
The first apply_function is overwritten and never called! @gemeinl 
So let's remove the OrderedDict and simply supply list. Another alternative as discussed before would be to call them line by line with some syntax like:

```python
transform_concat_ds(pick_types, eeg=True, meg=False, stim=False)
transform_concat_ds(apply_function, fun=lambda x: x * 1e6, channel_wise=False)
```

Which one you guys prefer? @hubertjb @sliwy @gemeinl

**Comment by robintibor** [comment]
Also, as we have confused @agramfort  ;) https://github.com/braindecode/braindecode/pull/90#issuecomment-593372271 maybe we can also think about different name for the function? Like `apply_to_raws` or `apply_to_concat_ds` ? Although I feel `transform_raws` would be fine for me (*ducking from @agramfort  ;)*)

**Comment by agramfort** [comment]
if you call it transforms everywhere stick to this word. Unless you want to
differentiate what apply does vs transform

>

**Comment by hubertjb** [comment]
@robintibor what would the list look like for the example you linked? I suppose something like this?

```python
raw_transforms = [ 
      ["pick_types", dict(eeg=True, meg=False, stim=False)], 
      ['apply_function', dict(fun=lambda x: x*1e6, channel_wise=False)], 
      ['filter', dict(l_freq=low_cut_hz, h_freq=high_cut_hz)], 
      ['apply_function', dict(fun=standardize_func, channel_wise=False)] 
 ]
```

**Comment by robintibor** [comment]
Yes like that @hubertjb

**Comment by robintibor** [comment]
I will have a try with that right now actually since it is giving unexpected results with dict. we can still see which interface we prefer after.

**Comment by hubertjb** [comment]
I'm personally in favor of the list approach then. I like the convenience of having all the transforms in a single variable that can be reused if need to.

**Comment by robintibor** [comment]
oki, now implemented in #94
